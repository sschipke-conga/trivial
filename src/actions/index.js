export const setTeams = (teams) => ({
  type: 'SET_TEAMS',
  teams
})

export const setQuestions = (questions) => ({
  type: 'SET_QUESTIONS',
  questions
})