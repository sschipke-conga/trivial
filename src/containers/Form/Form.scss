form {
  background: url(../../images/question-mark-background-for-powerpoint-12.jpg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  display: flex;
  align-items: flex-start;
  justify-content: center;
  padding: 20px;
  padding-top:50px; ;
  width: 100vw;
  height: 100vh;
}



input {
  font-size: 1.2em;
  padding: 10px;
  &:hover {
    cursor: text;
  }
}

.selection-div {
  width: 40%;
  .small {
  color: rgb(191, 78, 247);
  font-size: .7em;
}
.trivial {
  color: #FFB002;
  font-size: 2em;
}
}

select {
  font-size: 1.2em;
  &:hover {
    cursor: pointer;
  }
}

label {
  font-size: 1.2em;
  display: block;
  margin: 10px 0;
}

.team1-div,
.team2-div {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 45vh;
}

#teamOne, #teamTwo {
  font-size: 1.2em;
  width: 200px;
}
.selection-div {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: 10px;
  label, input, select {
    width: 60%;
  }
}

.play-button {
  background: #FFB002;
  color: white;
  font-size: 1.2em;
  height: 40px;
  margin-top: 40px;
  width: 60%;
  &:hover {
    color: black;
    box-shadow: 2px 2px 2px black;
  }
}