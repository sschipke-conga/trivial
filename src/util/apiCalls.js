export const fetchQuestions = async (amount, category, difficulty) => {
  let url = `https://opentdb.com/api.php?amount=${amount}&category=${category}&difficulty=${difficulty}&type=multiple`
  const res = await fetch(url);
  if (!res.ok) {
    throw Error('Failed to fetch')
  }
  const data = await res.json()
  return data.results
}
