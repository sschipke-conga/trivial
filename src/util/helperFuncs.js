export const shuffle = answers => {
  var m = answers.length, t, i;
  while (m) {
    i = Math.floor(Math.random() * m--);
    t = answers[m];
    answers[m] = answers[i];
    answers[i] = t;
  }
  return answers;
}

export const determineCurrentTeam = (turnCount) => {
  if (turnCount % 2 === 0) {
    return 1
  } 
  return 0
}

export const determineWinner = teams => {
  if(teams[0].score === teams[1].score) {
    return false
  }
  return teams.sort((teamOne, teamTwo) => {
    return teamTwo.score-teamOne.score
  })[0]
}

export const capDifficulty = difficulty => {
  return difficulty[0].toUpperCase() + difficulty.slice(1)
}

export const cleanEncoding = (string) => {
  let cleanPhrase = string.replace(/&#039;/g, "'")
  .replace(/&quot;/g, '"')
  .replace(/&eacute/g, 'Ã©');
  return cleanPhrase
}

export const cleanQuestions = (questions) => {
    return questions.map(question => {
    question.question = cleanEncoding(question.question)
    question.correct_answer = cleanEncoding(question.correct_answer)
    question.incorrect_answers = question.incorrect_answers.map(answer => cleanEncoding(answer))
    return question
    })
}