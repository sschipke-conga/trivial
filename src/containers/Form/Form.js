import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {setTeams, setQuestions, setHaveQuestions} from '../../actions/index';
import {fetchQuestions} from '../../util/apiCalls';
import './Form.scss';

export class Form extends Component {
  constructor() {
    super() 
    this.state = {
      teamOne:'',
      teamTwo: '',
      amount: '',
      category: '',
      difficulty: ''
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.id]:e.target.value})
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    this.handleQuestions()
    this.handleTeams()
  }

  handleTeams = () => {
    const { teamOne, teamTwo } = this.state;
    const {setTeams} = this.props;
    const teams = [{
      name: teamOne,
      score: 0
    }, { name: teamTwo, score: 0 }]
    setTeams(teams)
  }

  handleQuestions = async () => {
    const {amount, category, difficulty} = this.state;
    const {setQuestions, setHaveQuestions} = this.props;
    let questions = await fetchQuestions((amount*2), category, difficulty);
    setQuestions(questions)
    setHaveQuestions()
  }

  render() {
    const { teamOne, amount, category, difficulty, teamTwo } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <div className="team1-div">
          <label className="team-label" htmlFor="name-team1">Team 1 Name</label>
          <input type="text" id="teamOne" maxLength="20" value={teamOne}
          onChange={this.handleChange} placeholder="Team 1 name" required />
        </div>
        <div className="selection-div">
          <h1>Welcome to It's not <span>small</span> it's <span>Trivial</span>!</h1>
          <label htmlFor="amount">Choose the number of rounds (up to 10)</label>
          <input type="number" id="amount" min="1" max="10" value={amount}
          placeholder="number of rounds" onChange={this.handleChange} required/>
          <label htmlFor="category">Select a category</label>
          <select value={category}
            onChange={this.handleChange} id="category">
            <option value="9" defaultValue>General Knowledge</option>
            <option value="11">Movies</option>
            <option value="13">Musicals/Theatre</option>
            <option value="20">Mythology</option>
            <option value="23">History</option>
            <option value="17">Science & Nature</option>
          </select>
          <label htmlFor="category">Select the difficulty</label>
          <select value={difficulty}
            onChange={this.handleChange} id="difficulty">
            <option value="" defaultValue>Any</option>
            <option value="easy">Easy</option>
            <option value="medium">Medium</option>
            <option value="hard">Hard</option>
          </select>
          <button className="play-button" type="submit">Play!</button>
        </div>
        <div className="team2-div">
          <label htmlFor="name-team1">Team 1 Name</label>
          <input type="text" id="teamTwo" maxLength="20" 
          onChange={this.handleChange} value={teamTwo}
          placeholder="Team 2 name" required />
        </div>
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => 
  bindActionCreators({setTeams, setQuestions, setHaveQuestions}, dispatch)

export default connect(null, mapDispatchToProps)(Form);

Form.propTypes = {
  setTeams: PropTypes.func
}