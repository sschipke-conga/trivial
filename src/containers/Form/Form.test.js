import React from 'react';
import { Form, mapDispatchToProps } from './Form';
import { shallow } from 'enzyme';
import { fetchQuestions } from '../../util/apiCalls';
import * as actions from '../../actions';

jest.mock('../../util/apiCalls');
const mockQuestions = [{
  "category": "Sports",
  "type": "multiple",
  "difficulty": "medium",
  "question": "At which bridge does the annual Oxford and Cambridge boat race start?",
  "correct_answer": "Putney",
  "incorrect_answers": [
    "Hammersmith",
    "Vauxhall ",
    "Battersea"
  ]
},
{
  "category": "Entertainment: Video Games",
  "type": "multiple",
  "difficulty": "easy",
  "question": "Who is Sonic&#039;s sidekick?",
  "correct_answer": "Tails",
  "incorrect_answers": [
    "Shadow",
    "Amy",
    "Knuckles"
  ]
},
{
  "category": "Politics",
  "type": "boolean",
  "difficulty": "easy",
  "question": "There was a satirical candidate named &quot;Deez Nuts&quot; running in the 2016 US presidential elections.",
  "correct_answer": "True",
  "incorrect_answers": [
    "False"
  ]
}]

fetchQuestions.mockImplementation(() => 
  Promise.resolve(mockQuestions)
  )

const mockTeams = [{
  name: 'Lez Quizzerable',
  score: 0
}, {
  name: "Let's Get Quizzacle",
  score: 0
}]
actions.setTeams = jest.fn().mockImplementation(() => {
  return mockTeams
})

actions.setHaveQuestions = jest.fn()

actions.setQuestions = jest.fn()

describe('Form', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Form setTeams={actions.setTeams} setHaveQuestions={actions.setHaveQuestions} setQuestions={actions.setQuestions}/>
    )
  })
  it('should match the initial snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  describe('handleChange', () => {
    it('should update the state when team One types in their name', () => {
      const mockEvent = { target: { id: 'teamOne', value: 'Lez Quizzerable' } };
      expect(wrapper.state('teamOne')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('teamOne')).toEqual(mockEvent.target.value)
    })
    it('should update the state when team Two types in their name', () => {
      const mockEvent = { target: { id: 'teamTwo', value: "Let's Get Quizzacle" } };
      expect(wrapper.state('teamTwo')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('teamTwo')).toEqual(mockEvent.target.value)
    })
    it('should update the state when the user types in an amount', () => {
      const mockEvent = { target: { id: 'amount', value: '5'} };
      expect(wrapper.state('amount')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('amount')).toEqual(mockEvent.target.value)
    })
    it('should update the state when the user selects a category', () => {
      const mockEvent = { target: { id: 'category', value: '11' } };
      expect(wrapper.state('category')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('category')).toEqual(mockEvent.target.value)
    })
    it('should update the state when the user selects their difficulty', () => {
      const mockEvent = { target: { id: 'difficulty', value: 'medium' } };
      expect(wrapper.state('difficulty')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('difficulty')).toEqual(mockEvent.target.value)
    })
    it('should run handleChange when the inputs detect a change', () => {
      const mockTeamOneEvent = { target: { id: 'teamOne', value: 'Robbie' } };
      const mockTeamTwoEvent = { target: { id: 'teamTwo', value: 'Travis' } };
      wrapper.instance().handleChange = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('#teamOne').simulate('change', mockTeamOneEvent);
      wrapper.find('#teamTwo').simulate('change', mockTeamTwoEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockTeamOneEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockTeamTwoEvent);
    }) 
  })
  describe('handleSubmit', () => {
    it('should prevent the default action', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      wrapper.instance().handleSubmit(mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalled()
    })
    it('should call handleQuestions', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      wrapper.instance().handleQuestions = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().handleQuestions).toHaveBeenCalled()
    })
    it('should call handleTeamns', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      wrapper.instance().handleTeams = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().handleTeams).toHaveBeenCalled()
    })
    it('should call handleSubmit when the form is submitted', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().handleSubmit = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('form').simulate('submit', mockEvent);
      expect(wrapper.instance().handleSubmit).toHaveBeenCalledWith(mockEvent)
    })
  })
  describe('handleTeams', () => {
    it('should call setTeams', () => {
      wrapper.instance().setState({ teamOne: 'Lez Quizzerable', teamTwo: "Let's Get Quizzacle"})
      wrapper.instance().handleTeams();
      expect(actions.setTeams).toHaveBeenCalledWith(mockTeams)
    })
  })
  describe('handleQuestions', () => {
    it('should call fetchQuestions', () => {
      wrapper.instance().handleQuestions();
      expect(fetchQuestions).toHaveBeenCalled()
    })
    it('should call setQuestions with he correct questions', () => {
      wrapper.instance().handleQuestions()
      expect(actions.setQuestions).toHaveBeenCalledWith(mockQuestions)
    })
    it('should call setHaveQuestions', () => {
      wrapper.instance().handleQuestions()
      expect(actions.setHaveQuestions).toHaveBeenCalled()
    })
  })
  describe('mapDispatchToProps', () => {
    it('should call Dispatch with setTeams actions when setTeams is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setTeams('SET_TEAMS', mockTeams)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setTeams('SET_TEAMS', mockTeams)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
    it('should call Dispatch with setQuestions actions when setQuestions is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setQuestions('SET_QUESTIONS', mockQuestions)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setQuestions('SET_QUESTIONS', mockQuestions)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
    it('should call Dispatch with setHaveQuestions actions when setHaveQuestions is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setHaveQuestions('SET_HAVE_QUESTIONS')
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setHaveQuestions('SET_HAVE_QUESTIONS')
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})