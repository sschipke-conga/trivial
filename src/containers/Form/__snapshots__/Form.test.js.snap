// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the initial snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="team1-div"
  >
    <label
      className="team-label"
      htmlFor="name-team1"
    >
      Name of team one
    </label>
    <input
      id="teamOne"
      maxLength="20"
      onChange={[Function]}
      placeholder="Team 1 name"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="selection-div"
  >
    <h1>
      Welcome to 
    </h1>
    <h1>
      It's not 
      <span
        className="small"
      >
        small
      </span>
      , it's 
      <span
        className="trivial"
      >
        Trivial!
      </span>
    </h1>
    <label
      htmlFor="amount"
    >
      Number of rounds (up to 10)
    </label>
    <input
      id="amount"
      max="10"
      min="1"
      onChange={[Function]}
      placeholder="Number of rounds"
      required={true}
      type="number"
      value=""
    />
    <label
      htmlFor="category"
    >
      Select a category
    </label>
    <select
      id="category"
      onChange={[Function]}
      value=""
    >
      <option
        defaultValue={true}
        value="9"
      >
        General Knowledge
      </option>
      <option
        value="11"
      >
        Movies
      </option>
      <option
        value="13"
      >
        Musicals/Theatre
      </option>
      <option
        value="20"
      >
        Mythology
      </option>
      <option
        value="23"
      >
        History
      </option>
      <option
        value="17"
      >
        Science & Nature
      </option>
    </select>
    <label
      htmlFor="category"
    >
      Select the difficulty
    </label>
    <select
      id="difficulty"
      onChange={[Function]}
      value=""
    >
      <option
        defaultValue={true}
        value=""
      >
        Any
      </option>
      <option
        value="easy"
      >
        Easy
      </option>
      <option
        value="medium"
      >
        Medium
      </option>
      <option
        value="hard"
      >
        Hard
      </option>
    </select>
    <button
      className="play-button"
      type="submit"
    >
      Play!
    </button>
  </div>
  <div
    className="team2-div"
  >
    <label
      htmlFor="name-team1"
    >
      Name of team two
    </label>
    <input
      id="teamTwo"
      maxLength="20"
      onChange={[Function]}
      placeholder="Team 2 name"
      required={true}
      type="text"
      value=""
    />
  </div>
</form>
`;
